### PR TITLE
DOC: dot: Add explanation in case `b` has only 1 dimension.

### PR DESCRIPTION
**Pull request by xuhdev**

### CODE REVIEW

**Review by eric-wieser — COMMENTED** [review]
Commit message should start with ``DOC: `` too

File: numpy/add_newdocs.py [hunk 1]
```diff
@@ -1933,7 +1933,8 @@ def luf(lamdaexpr, *args, **kwargs):
     For 2-D arrays it is equivalent to matrix multiplication, and for 1-D
     arrays to inner product of vectors (without complex conjugation). For
     N dimensions it is a sum product over the last axis of `a` and
-    the second-to-last of `b`::
+    the second-to-last (or the only axis if `b` has only one dimension) of
+    `b`::
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: Perhaps ``second-to-last (or only, if `b` is 1D) axis`` would be clearer?

**Comment by xuhdev** [comment]
@eric-wieser Updated, thanks!

File: numpy/add_newdocs.py [hunk 2]
```diff
@@ -1933,7 +1933,7 @@ def luf(lamdaexpr, *args, **kwargs):
     For 2-D arrays it is equivalent to matrix multiplication, and for 1-D
     arrays to inner product of vectors (without complex conjugation). For
     N dimensions it is a sum product over the last axis of `a` and
-    the second-to-last of `b`::
+    the second-to-last (or only, if `b` is 1-D) of `b`::
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: I think you need the word `axis` after the parens here

**Comment by xuhdev** [comment]
Updated, Thanks!

File: numpy/add_newdocs.py [hunk 3]
```diff
@@ -1933,7 +1933,7 @@ def luf(lamdaexpr, *args, **kwargs):
     For 2-D arrays it is equivalent to matrix multiplication, and for 1-D
     arrays to inner product of vectors (without complex conjugation). For
     N dimensions it is a sum product over the last axis of `a` and
-    the second-to-last of `b`::
+    the second-to-last (or only, if `b` is 1-D) axis of `b`::
```
Comment (charris) [inline numpy/add_newdocs.py]: Hmm, this still doesn't capture everything, and parenthesis should be avoided here. Maybe a list of possibilities: matrix-matrix, matrix-vector, vector-vector, scalar. Or maybe start with `ndim >= 2` -- stacked matrices, `ndim == 1` -- vector, `ndim == 0` -- scalar, and then make a list of combinations as below

**Comment by xuhdev** [comment]
Updated again, thanks!

**Review by eric-wieser — COMMENTED** [review]
Looking pretty good now!

File: numpy/add_newdocs.py [hunk 4]
```diff
@@ -1928,12 +1928,19 @@ def luf(lamdaexpr, *args, **kwargs):
     """
     dot(a, b, out=None)
 
-    Dot product of two arrays.
+    Dot product of two arrays. Specifically,
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: One more case needed - if either of `a` or `b` are 0-D/scalars, then elementwise multiplication is used (`np.multiply`)

**Comment by xuhdev** [comment]
Fixed, thanks!

**Comment by xuhdev** [comment]
Any plan to merge this? :)

File: numpy/add_newdocs.py [hunk 5]
```diff
@@ -1928,12 +1928,22 @@ def luf(lamdaexpr, *args, **kwargs):
...
+
+    - If both `a` and `b` are 2-D arrays, it is matrix multiplication.
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: Perhaps add some mention of ``` ``a @ b`` ``` here?

**Comment by eric-wieser** [comment]
Was actually looking for this last night, but couldn't find it because it didn't have the label! Looks great and strictly better than before, although a brief mention of `@` would be nice, as commented above, if you're willing to make one more revision

File: numpy/add_newdocs.py [hunk 6]
```diff
@@ -1928,12 +1928,22 @@ def luf(lamdaexpr, *args, **kwargs):
...
+    - If `a` is an N-D array and `b` is a 1-D array, it is a sum product over
+      the last axis of `a` and `b`.
 
-    For 2-D arrays it is equivalent to matrix multiplication, and for 1-D
-    arrays to inner product of vectors (without complex conjugation). For
-    N dimensions it is a sum product over the last axis of `a` and
-    the second-to-last of `b`::
+    - If `a` is an N-D array and `b` is an M-D array (where ``M>=2``), it is a
+      sum product over the last axis of `a` and the second-to-last axis of `b`,
+      e.g.,::
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: Nit: `e.g.` seems slightly unnecessary in the presence of the `:` in the output.

**Comment by xuhdev** [comment]
@eric-wieser  Updated again. Thanks!

File: numpy/add_newdocs.py (hunk 4 above)
Comment (shoyer) [reply numpy/add_newdocs.py]: Is this behavior that we want to legitimitize by documenting?

Obviously, we won't change it at this point, but I would still encourage users to use `*` over dot in every circumstance.

**Review by eric-wieser — APPROVED** [review]

File: numpy/add_newdocs.py (hunk 4 above)
Comment (eric-wieser) [reply numpy/add_newdocs.py]: @shoyer: Isn't that true of this whole list? Isn't `*` or `@` always preferable over dot?

**Comment by eric-wieser** [comment]
If you put `[ci-skip]` on the last line of your commit message, then it won't trigger the tests to run - which for documentation changes, is what you want.

Either way, l'll leave this until @shoyer is happy

**Comment by charris** [comment]
@shoyer Ping.

File: numpy/add_newdocs.py (hunk 4 above)
Comment (shoyer) [reply numpy/add_newdocs.py]: Hmm. Can we explicitly recommend using matmul or times instead of dot in the docstring here then?

Comment (xuhdev) [reply numpy/add_newdocs.py]: @shoyer Please see my updates. Thanks!

File: numpy/add_newdocs.py [hunk 7]
```diff
@@ -1928,12 +1928,22 @@ def luf(lamdaexpr, *args, **kwargs):
...
+    Dot product of two arrays. Specifically,
+
+    - If both `a` and `b` are 1-D arrays, it is inner product of vectors
+      (without complex conjugation).
+
+    - If both `a` and `b` are 2-D arrays, it is matrix multiplication,
+      but using :func:`matmul` or ``a @ b`` is more preferred.
+
+    - If either `a` or `b` is 0-D (scalar), it is equivalent to :func:`multiply`
+      but using ``numpy.multiply(a, b)`` or ``a * b`` is more preferred.
```
Comment (charris) [inline numpy/add_newdocs.py]: ```and using ``numpy.multiply(a, b)`` or ``a * b`` is preferred.```

Comment (xuhdev) [reply numpy/add_newdocs.py]: @charris Fixed.

**Comment by charris** [comment]
Thanks @xuhdev .

**Pull request merged**